resources/regexes.ts

```diff
@@ -131,26 +203,88 @@ const parseHelper = <T extends LogDefinitionTypes>(
+    if (typeof parseField !== 'object')
+      throw new Error(`${defKey}: invalid value: ${JSON.stringify(parseField)}`);
 
     const fieldDefault = fieldName !== undefined && fieldsWithPotentialColons.includes(fieldName)
       ? matchWithColonsDefault
       : matchDefault;
-    const fieldValue = fields[keyStr]?.value?.toString() ?? fieldDefault;
-
-    if (fieldName !== undefined) {
-      str += Regexes.maybeCapture(
-        // more accurate type instead of `as` cast
-        // maybe this function needs a refactoring
-        capture,
-        fieldName,
-        params[fieldName],
-        fieldValue,
-      );
+    const defaultFieldValue = parseField.value?.toString() ?? fieldDefault;
+    const fieldValue = params[fieldName];
+
+    if (isRepeatingField(fields[keyStr]?.repeating, fieldValue)) {
+      let repeatingArray: RepeatingFieldsMap<T> = fieldValue;
+      const origRepeatingArray = repeatingArray;
+
+      const sortFn = fields[keyStr]?.sortFn;
+      // Allow sorting if needed
+      if (sortFn)
+        repeatingArray = [...repeatingArray].sort(sortFn);
+
+      repeatingArray.forEach((rep) => {
+        if (typeof rep === 'object' && rep !== null && Object.keys(rep).length > 0) {
+          let fieldRegex = '';
+          // Rather than looping over the keys defined on the object,
+          // loop over the base type def's keys. This enforces the correct order.
+          fields[keyStr]?.repeatingKeys?.forEach((key) => {
+            fieldRegex += separator;
+            if (!(key in rep)) {
+              // If we don't have a value for this key, insert a placeholder
+              fieldRegex += Regexes.maybeCapture(
+                capture,
+                // All capturing groups have their index appended
+                key + origRepeatingArray.indexOf(rep).toString(),
+                defaultFieldValue,
+                defaultFieldValue,
+              );
+              return;
+            }
+            // This is an ugly cast. There's probably a better way of narrowing this type but
+            // tsc doesn't like the dozen or so variations I've tried.
+            const val = rep[key as keyof typeof rep];
+            if (typeof val !== 'string') {
+              if (!Array.isArray(val))
+                return;
+              if (val.length < 0)
+                return;
+              if (val.some((v) => typeof v !== 'string'))
+                return;
+            }
+            fieldRegex += Regexes.maybeCapture(
+              capture,
+              // All capturing groups have their index appended
+              key + origRepeatingArray.indexOf(rep).toString(),
+              val,
+              defaultFieldValue,
+            );
+          });
+
+          if (fieldRegex.length > 0) {
+            // Prepend a match for any number of fields before matching this one.
+            // TODO: Should we scope this as the following instead?
```

Comment:
Yeah, I think probably we should.  Probably it's the case that no field name is going to match a field value, but that would be a really hairy bug to sort out later.